core/runtime/src/main/java/io/quarkus/runtime/execution/ExecutionChain.java

```diff
@@ -0,0 +1,193 @@
+package io.quarkus.runtime.execution;
+
+import java.io.Flushable;
+import java.io.Serializable;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.concurrent.ExecutionException;
+
+import org.wildfly.common.Assert;
+
+import com.oracle.svm.core.annotate.AlwaysInline;
+
+import io.quarkus.runtime.Timing;
+
+/**
+ * Application startup chain node.
+ */
+public final class ExecutionChain implements ExecutionHandler, Serializable {
+    private static final long serialVersionUID = 7360468681846082188L;
+
+    public static final String QUARKUS_FEATURES = "quarkus.features";
+    public static final String QUARKUS_VERSION = "quarkus.version";
+
+    private final ExecutionChain next;
+    private final ExecutionHandler handler;
+
+    /**
+     * Construct a new instance. This should generally be done by the container during static initialization, not by
+     * the user.
+     *
+     * @param next the next item in the chain, or {@code null} to indicate the end of the chain
+     * @param handler the corresponding startup handler (must not be {@code null})
+     */
+    public ExecutionChain(final ExecutionChain next, final ExecutionHandler handler) {
+        Assert.checkNotNullParam("handler", handler);
+        this.next = next;
+        this.handler = handler;
+    }
+
+    /**
+     * Pass control to the next startup handler. This method may block indefinitely, and will only return upon ordered
+     * application shutdown.
+     *
+     * @param context the startup context to pass to the next step (must not be {@code null})
+     * @return the exit code
+     * @throws Exception if the next startup handler fails
+     */
+    @AlwaysInline("Application startup")
+    public int proceed(ExecutionContext context) throws Exception {
+        Assert.checkNotNullParam("context", context);
+        if (next != null) {
+            return next.run(context);
```

Comment:
> * This model makes adding parallel startup much more difficult, as it requires all startup takes to be run in the same stack

There is nothing stopping startup tasks from distributing their work across multiple threads, as long as they occur after the thread pool is set up.

> * If there are lots of steps this can overflow the stack (i.e. the same issue we had when EAP used this approach)

Even if every extension had 10 steps, and each step used 200 bytes of stack, you'd need thousands of active extensions in one application to even come close to hitting this problem.  Do you really believe this is a risk?

> * It can be confusing to the users when there is a large stack trace present in a thread dump

How so?

Reply:
> > * This model makes adding parallel startup much more difficult, as it requires all startup takes to be run in the same stack
> 
> There is nothing stopping startup tasks from distributing their work across multiple threads, as long as they occur after the thread pool is set up.
> 
> > * If there are lots of steps this can overflow the stack (i.e. the same issue we had when EAP used this approach)
> 
> Even if every extension had 10 steps, and each step used 200 bytes of stack, you'd need thousands of active extensions in one application to even come close to hitting this problem. Do you really believe this is a risk?

Probably not, but I also just don't really see what the advantage is?

> 
> > * It can be confusing to the users when there is a large stack trace present in a thread dump
> 
> How so?

Because if you don't know anything about Quarkus and you see a large stack trace you will likely assume it is related to your application and start looking there. IDE's also generally sort by stack depth in their snapshots, so this thread will often be the first thing a user sees.

Reply:
> > > * This model makes adding parallel startup much more difficult, as it requires all startup takes to be run in the same stack
> > 
> > 
> > There is nothing stopping startup tasks from distributing their work across multiple threads, as long as they occur after the thread pool is set up.
> > > * If there are lots of steps this can overflow the stack (i.e. the same issue we had when EAP used this approach)
> > 
> > 
> > Even if every extension had 10 steps, and each step used 200 bytes of stack, you'd need thousands of active extensions in one application to even come close to hitting this problem. Do you really believe this is a risk?
> 
> Probably not, but I also just don't really see what the advantage is?

Well, if you don't see it by now, then perhaps you never will, but I think it might be more apparent as the next phases of this come in.

> > > * It can be confusing to the users when there is a large stack trace present in a thread dump
> > 
> > 
> > How so?
> 
> Because if you don't know anything about Quarkus and you see a large stack trace you will likely assume it is related to your application and start looking there. IDE's also generally sort by stack depth in their snapshots, so this thread will often be the first thing a user sees.

It *is* related to the application though.  I don't really see what the problem is.